### PR TITLE
Create admin page & backend overhaul

### LINKED ISSUE #107: Create admin page
Create page used by admins and portfolio managers to define document types/document classes/whatever it's called. Will allow users to add, edit, and remove document classes with functionality for names, descriptions, and examples (not ready on backend but will implement here in frontend).

### PR DESCRIPTION
**Pull request by arpanlaha**
Resolves #107 
So when I was working through the Document Class page I found that a lot of the functionality required pretty significant changes to how our backend works. Rather than make a few of these changes and make our backend extremely consistent, I opted to overhaul the backend at the cost of this PR becoming absolutely disgusting. I've heavily documented my changes to try to make this easier to understand, but ask any questions you need to.

### CODE REVIEW

File: backend/api/__init__.py [hunk 1]
```diff
@@ -78,10 +78,11 @@ def create_app(test_config=None):
     Migrate(app, db)
 
     # import and register blueprints
-    from api.views import main, message, box, fp, pm, docclass
+    from api.views import main, document, message, box, fp, pm, docclass
```
Comment (arpanlaha) [inline backend/api/__init__.py]: I moved all Document views into their own file for clarity and consistency with other tables

File: backend/api/models/DocumentClass.py [hunk 2]
```diff
@@ -11,11 +11,16 @@ class DocumentClass(Mixin, db.Model):
     id = db.Column(db.String, unique=True, primary_key=True)
     name = db.Column(db.String, unique=True)
     description = db.Column(db.String, unique=False, nullable=True)
+    example = db.Column(db.String, unique=False, nullable=True)
+    # an example of this Document Class, represented by a shared Box link
 
-    def __init__(self, name, description=None):
+    def __init__(self, data):
```
Comment (arpanlaha) [inline backend/api/models/DocumentClass.py]: I changed the `__init__()` for all models from taking in parameters individually to taking in a dictionary containing the parameters as a result of the shift to FormData for all POST requests (covered later). If the method was left as is, then some parameters would be returned as `"name": ["Angad"]` instead of `"name": "Angad"`, or similar.

File: backend/api/models/FieldPartner.py [hunk 3]
```diff
@@ -15,12 +15,12 @@ class FieldPartner(Mixin, db.Model):
     pm_id = db.Column(db.String, db.ForeignKey("portfolio_manager.id"))
     app_status = db.Column(db.String)
 
-    def __init__(self, email, org_name, pm_id, app_status):
+    def __init__(self, data):
```
Comment (arpanlaha) [inline backend/api/models/FieldPartner.py]: See above

File: backend/api/models/PortfolioManager.py [hunk 4]
```diff
@@ -13,13 +13,11 @@ class PortfolioManager(Mixin, db.Model):
     id = db.Column(db.String, unique=True, primary_key=True)
     email = db.Column(db.String)
     name = db.Column(db.String)
-    list_of_fps = db.Column(ARRAY(db.String), nullable=True)
 
-    def __init__(self, email, name, list_of_fps):
+    def __init__(self, data):
```
Comment (arpanlaha) [inline backend/api/models/PortfolioManager.py]: See above

File: backend/api/views/box.py [hunk 5]
```diff
@@ -189,7 +189,15 @@ def get_file_info(client, file_id):
 
 @box.route("/box/download", methods=["GET"])
 def download_file():
-    file_id = request.form.get("file_id")
+    data = request.form
```
Comment (arpanlaha) [inline backend/api/views/box.py]: I just added this stuff for consistency with all other routes.

File: backend/api/views/docclass.py [hunk 6]
```diff
@@ -26,28 +27,63 @@ def get_document_class_by_id(id):
 @docclass.route("/document_class/new", methods=["POST"])
 def add_document_class():
     """ function that is called when you visit /document_class/new, creates a new docclass """
-    data = request.get_json()
-    logger.info(data)
+    data = request.form
```
Comment (arpanlaha) [inline backend/api/views/docclass.py]: This PR changes the body through which parameters are passed to PUT and POST requests from `json` to `FormData`. Any request that handles files requires that we use FormData; however using `request.json` or `request.get_json()` on these does not work. For consistency's sake, and because we can get the proper results with FormData provided we changed the model `__init__()`s as above, I changed any instance of `request.json` or `request.get_json()` to `request.form`.

File: backend/api/views/document.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
+from flask import Blueprint, request, json
```
Comment (arpanlaha) [inline backend/api/views/document.py]: New file (see above).

File: backend/api/views/fp.py [hunk 8]
```diff
@@ -64,15 +64,15 @@ def new_fp():
             status=400, message="No application status provided for new FP"
         )
     sample_args = request.args
-    new_fp = FieldPartner(**data)
+    new_fp = FieldPartner(data)
```
Comment (arpanlaha) [inline backend/api/views/fp.py]: As a result of the new `__init()__` syntax, we can simply pass in `data` to model constructors instead of using `**data`.

File: backend/api/views/main.py [hunk 9]
```diff
@@ -1,11 +1,7 @@
-from flask import Blueprint, request, json
```
Comment (arpanlaha) [inline backend/api/views/main.py]: Remove anything pertaining specifically to Documents (see above).

File: backend/api/views/pm.py [hunk 10]
```diff
@@ -34,34 +34,19 @@ def get_pm_by_email(email):
     )
 
 
-@pm.route("/portfolio_manager/all_fps/<id>", methods=["GET"])
```
Comment (arpanlaha) [inline backend/api/views/pm.py]: List of Field Partners removed from Portfolio Manager table - see model file for elaboration.

File: backend/tests/test_document.py [hunk 11]
```diff
@@ -76,115 +76,6 @@ def test_get_document(client):
     assert ret_dict["result"]["documents"]["Pending"][0]["status"] == "Pending"
 
 
-"""
```
Comment (arpanlaha) [inline backend/tests/test_document.py]: Remove old comment (why was this here?, I'm a bit confused).

File: backend/tests/test_document.py [hunk 12]
```diff
@@ -205,9 +96,9 @@ def test_update_status(client):
     db.session.commit()
 
     rs = client.put(
-        "/document/status",
-        content_type="application/json",
-        json={"status": "Missing", "docID": temp_document.id},
+        "/document/status/" + str(temp_document.id),
+        content_type="multipart/form-data",
+        data={"status": "Missing"},
```
Comment (arpanlaha) [inline backend/tests/test_document.py]: As a result of using FormData, our test syntax needs to change. Instead of `content_type="application/json`, we need to use `content_type="multipart/form-data"`, and instead of `json={...}`, we need to use `data={...}`.

File: backend/api/models/PortfolioManager.py [hunk 13]
```diff
@@ -13,13 +13,11 @@ class PortfolioManager(Mixin, db.Model):
...
     email = db.Column(db.String)
     name = db.Column(db.String)
-    list_of_fps = db.Column(ARRAY(db.String), nullable=True)
 
-    def __init__(self, email, name, list_of_fps):
+    def __init__(self, data):
         self.id = "p" + str(uuid.uuid4())
-        self.email = email
-        self.name = name
-        self.list_of_fps = list_of_fps
```
Comment (arpanlaha) [inline backend/api/models/PortfolioManager.py]: We shoudn't be storing things in arrays with Postgres, and since we already store a Porfolio Manager ID with each Field Partner, we can simply query the Field Partner table to find all the Field Partners for a given Portfolio Manager.

File: frontend/src/components/DocumentClass.js [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+              <span> Description: </span>
+              <textarea
+                name="paragraph_text"
+                cols="50"
+                rows="10"
+                onChange={this.updateDescription}
+                value={this.state.description}
+              />
+              <br />
+              <div className="dropPage">
```
Comment (arpanlaha) [inline frontend/src/components/DocumentClass.js]: Ripped from `Upload` but modified for different backend interactions.

File: frontend/src/components/DocumentClassPage.js [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+              <br />
+              <span> Description: </span>
+              <textarea
+                name="paragraph_text"
+                cols="50"
+                rows="10"
+                onChange={this.updateDescription}
+              />
+              <br />
+              <div className="dropPage">
```
Comment (arpanlaha) [inline frontend/src/components/DocumentClassPage.js]: Same as above.

File: frontend/src/components/Upload.js [hunk 16]
```diff
@@ -1,7 +1,7 @@
 import React, { Component } from 'react'
 import Dropzone from 'react-dropzone'
 import { Button, Modal, ModalBody, ModalFooter } from 'reactstrap'
-import { sendFile } from '../utils/ApiWrapper'
+import { uploadDocument } from '../utils/ApiWrapper'
```
Comment (arpanlaha) [inline frontend/src/components/Upload.js]: Renamed to make distinction between uploading Documents and Document Classes.

File: frontend/src/index.js [hunk 17]
```diff
@@ -21,11 +22,13 @@ ReactDOM.render(
   <Provider store={store}>
     <ConnectedRouter onUpdate={() => window.scrollTo(0, 0)} history={history}>
       <Switch>
-        <Route exact path="/" component={LogIn} />
```
Comment (arpanlaha) [inline frontend/src/index.js]: Not connected yet, lets stick to the old one for now.

File: frontend/src/components/DocumentClassPage.js [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+                    </ul>
+                    <Button
+                      disabled={this.state.files.length === 0}
+                      className="right"
+                      onClick={this.handleSubmit}
+                    >
+                      Create Document Class
+                    </Button>
+                    <Modal isOpen={this.state.uploadModal} toggle={this.secondToggle}>
+                      <ModalBody>File uploaded - your submission is being processed.</ModalBody>
```
Comment (rgychiu) [inline frontend/src/components/DocumentClassPage.js]: Im wondering if there's a need for nested modals - you might be able to just have field under one modal (based on just quickly running the application. lmk tho)

File: backend/api/models/FieldPartner.py [hunk 19]
```diff
@@ -15,12 +15,12 @@ class FieldPartner(Mixin, db.Model):
...
         self.id = "f" + str(uuid.uuid4())
-        self.email = email
-        self.org_name = org_name
-        self.pm_id = pm_id
-        self.app_status = app_status
+        self.email = data["email"]
+        self.org_name = data["org_name"]
+        self.pm_id = data["pm_id"]
+        self.app_status = data["app_status"]
```
[the remark this answers is not in the document]
Comment (arpanlaha) [reply backend/api/models/FieldPartner.py]: This should be done in the request itself, the reason I check the others is that nullable is set to true whereas these aren't nullable and as such should have been handled in the request.

File: backend/api/views/docclass.py [hunk 20]
```diff
@@ -26,28 +27,63 @@ def get_document_class_by_id(id):
...
     db.session.commit()
     return create_response(status=200, data={"document_class": updated_docclass})
+
+
+@docclass.route("/document_class/delete/<id>", methods=["DELETE"])
+def delete_document_class(id):
+    # delete all associated Documents before deleting Document Class
+    Document.query.filter(Document.docClassID == str(id)).delete()
+
+    db.session.delete(DocumentClass.query.filter((DocumentClass.id == str(id))).first())
```
[the remark this answers is not in the document]
Comment (arpanlaha) [reply backend/api/views/docclass.py]: Pretty sure it doesn't but I'll check

File: frontend/src/components/DocumentClassPage.js (hunk 18 above)
Comment (arpanlaha) [reply frontend/src/components/DocumentClassPage.js]: What do you mean by field?

File: backend/api/models/FieldPartner.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rgychiu) [reply backend/api/models/FieldPartner.py]: That makes sense, in that case then I think it's okay to leave it, but I would recommend leaving a comment that states that existence is checked elsewhere and that it's assumed that the fields exist when they hit that constructor/making a new entry and whatnot

File: frontend/src/components/DocumentClassPage.js (hunk 18 above)
Comment (rgychiu) [reply frontend/src/components/DocumentClassPage.js]: I think I meant everything, not quite sure what i meant there 8 hours ago lol

Comment (arpanlaha) [reply frontend/src/components/DocumentClassPage.js]: I changed it to just close the modal on submission instead of opening a new one
